File: src/main/java/net/miarma/mkernel/commands/misc/SendCoordsCommand.java
package net.miarma.mkernel.commands.misc;

import java.util.List;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Player;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.config.providers.MessageProvider;
import net.miarma.mkernel.util.MessageUtil;

public class SendCoordsCommand {
    public static void register() {
        CommandWrapper sendCoordsCmd = CommandProvider.getSendCoordsCommand();
        new CommandAPICommand(sendCoordsCmd.getName())
            .withArguments(CommandProvider.Arguments.playerArg())
            .withFullDescription(sendCoordsCmd.getDescription())
            .withPermission(sendCoordsCmd.getPermission().base())
            .withShortDescription(sendCoordsCmd.getDescription())
            .executesPlayer((sender, args) -> {
                if (args.count() > 1) {
                    MessageUtil.sendMessage(sender, MessageProvider.Errors.tooManyArguments(), true);
                }

                Player player = Bukkit.getPlayer(args.getRaw(0));
                Location loc = sender.getLocation();
                List<String> coords = List.of(String.valueOf(loc.getBlockX()),
                                            String.valueOf(loc.getBlockY()),
                                            String.valueOf(loc.getBlockZ()));

                MessageUtil.sendMessage(sender, sendCoordsCmd.getMessages()[0], true,
                                            List.of("%target%"), List.of(player.getName()));
                MessageUtil.sendMessage(player, sendCoordsCmd.getMessages()[1], true,
                                            List.of("%sender%", "%x%", "%y%", "%z%"),
                                            List.of(sender.getName(), coords.get(0), coords.get(1), coords.get(2)));
            })
            .register();
    }
}

File: src/main/kotlin/net/miarma/mkernel/command/impl/misc/SendCoordsCommand.kt
package net.miarma.mkernel.command.impl.misc

import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.arguments.PlayerProfileArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.command.MCommand
import net.miarma.mkernel.common.config.ConfigKeys
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.MessageService
import net.miarma.mkernel.util.PlayerUtil

@Singleton
class SendCoordsCommand @Inject constructor(
    private val configService: ConfigService,
    private val messageService: MessageService
) : MCommand {
    override fun register() {
        commandAPICommand(configService.getString(ConfigKeys.Commands.SendCoords.NAME)) {
            withArguments(PlayerProfileArgument(configService.getString(ConfigKeys.Arguments.PLAYER)))
            withFullDescription(configService.getString(ConfigKeys.Commands.SendCoords.DESC))
            withPermission(configService.getString(ConfigKeys.Commands.SendCoords.PERM))
            withUsage(configService.getString(ConfigKeys.Commands.SendCoords.USAGE))
            playerExecutor { sender, args ->
                val target = PlayerUtil.fromArg(args[0]) ?: run {
                    messageService.builder(configService.getString(ConfigKeys.Messages.General.Errors.PLAYER_NOT_FOUND)).withPrefix().send(sender)
                    return@playerExecutor
                }

                val loc = sender.location
                messageService.builder(configService.getString(ConfigKeys.Commands.SendCoords.MSG_SENT)).withPrefix().tag("target", target.name).send(sender)
                messageService.builder(configService.getString(ConfigKeys.Commands.SendCoords.MSG_COORDS))
                    .withPrefix().forPlayer(target).tag("sender", sender.name)
                    .tag("x", loc.blockX.toString()).tag("y", loc.blockY.toString()).tag("z", loc.blockZ.toString())
                    .send(target)
            }
        }
    }
}